### ISSUE TITLE
help for including odd orders during basex inversion

### ISSUE THREAD

**Issue by haribo3127** [PyAbel/PyAbel#265]
Hi,

I have a left-right symmetric images, but with positive values in the north hemisphere and negative values in the south hemisphere. I would like to force the use of odd orders for the inversion.

With the basex method, I can reconstruct such images but it is not possible to extract odd parameters. Is there a way to do so with this method? For example, linbasex allows to set legendre_order[1, 3].

Thanks!

**Comment by MikhailRyazanov** [comment]
I actually want to add odd orders to `abel.tools.vmi.Distributions`, but I need to know what is the most reasonable convention for them.

I saw some people using just a finite number of odd Legendre polynomials and some other people using even Legendre polynomials multiplied by cos θ (which is essentially equivalent to using odd polynomials, but with different conventions). And my impression is that in most physically meaningful cases, asymmetric distributions cannot be described _exactly_ by a finite number of odd orders (in contrast to many cases where even polynomials up to some predetermined order give the exact description), although I did not see any good literature about this.

Can you please explain what exactly you need and why (like what is the mathematical model of your angular distributions)? References to what _other_ people might need will be also very useful.

**Comment by haribo3127** [comment]
Sorry for my late reply. Indeed there seems to be a very few cases where the use of odd orders is meaningful. One of them is regarding photoelectron distributions of optically active (chiral) system. In this case one typically look at differential images where negative values can arise for one of the hemisphere.

I would refer to the following two references for the formalism, the first one gives a general overview for applying higher odd orders to the inversion procedure.
 
1. https://aip.scitation.org/doi/full/10.1063/1.4844295 (Section II)
2. https://journals.aps.org/pra/abstract/10.1103/PhysRevA.13.1411 (Original work)

In those cases, one already has inital information from the physical system to define the orders to be used for the reconstruction. Am I clear enough?

**Comment by haribo3127** [comment]
as an additional comment:
from the 'classResults(r, cn)' one see that beta0 = 1 by definition (does this mean that each beta parameters are normalized by beta0?). As some information can also be contained in beta0 does it makes sense to relax this constraint in the abel.tools.vmi.Distributions?

**Comment by MikhailRyazanov** [comment]
Thanks for the references! It seems to make sense that non-oriented chiral systems under circularly polarized light might produce a finite number of odd terms, although I did not read into details yet to see why higher-order terms vanish (and Janssen's group work does not seem to explain whether they have actually derived that multiphoton formula or just extended it by analogy).

In any case, adding odd terms would not harm (since they all are orthogonal, _P_<sub>1</sub>(cos θ) = cos θ still gives the total antisymmetric contribution). My concern was that if there are more appropriate descriptions, we should implement them. I'll try to take a look at implementing odd orders towards the end of the week...

Regarding β<sub>0</sub> = 1, this is just how the anisotropy parameters are defined. [This representation](https://pyabel.readthedocs.io/en/latest/abel.html#abel.tools.vmi.Distributions.Results.Ibeta) simply separates the total intensity _I_ from the angular dependences β<sub>_n_</sub>_P_<sub>_n_</sub>(cos θ). So since the 0-odrer term gives the total intensity, its β<sub>0</sub> cannot vary. In other words, it's not that “each beta parameters are normalized by beta0”, but actually each β<sub>_n_</sub> is defined as the _n_-th harmonic intensity normalized by the total intensity. In you want the harmonic intensities directly, without normalization, they are available as [`harmonics()`](https://pyabel.readthedocs.io/en/latest/abel.html#abel.tools.vmi.Distributions.Results.harmonics) (this might be helpful if you try to fit differential images with zero total intensity, since the usual normalization will not work).

**Comment by haribo3127** [comment]
thanks for the clarification! I'm looking forward to test the inversion with odd orders

**Comment by MikhailRyazanov** [comment]
Now I have the odd orders more or less working, but I need to finish testing them. @haribo3127, if you can provide a typical (or interesting) example of real data, it will also help.

**Comment by haribo3127** [comment]
Great! I will have a try as soon as next week and will look into some data to provide a test experimental image. A question arising from my reading of #266 : is the option to manually select the orders available (such as for the linbasex method)? For example, choosing to have only order 1 and 3 with no beta 0, 2 or 4?

**Comment by MikhailRyazanov** [comment]
So you have a purely antisymmetric image? Looks strange.

I mean, if it comes from VMI data, this should be a difference of two non-negative images, let's call them IM<sup>+</sup> and IM<sup>−</sup>, both of which should have at least the 0-order even term. And if you indeed need the anisotropy parameters β<sub>_n_</sub>, you will need these 0-order terms, since the anisotropy parameters are defined as β<sub>_n_</sub> = _I_<sub>_n_</sub> / _I_<sub>0</sub>. So I would suggest to analyze IM<sup>+</sup> and IM<sup>−</sup> separately using `order=3` and then calculate either 2β<sub>1</sub> = β<sub>1</sub><sup>+</sup> − β<sub>1</sub><sup>−</sup> or β<sub>1</sub> = (_I_<sub>1</sub><sup>+</sup> − _I_<sub>1</sub><sup>−</sup>) / (_I_<sub>0</sub><sup>+</sup> + _I_<sub>0</sub><sup>−</sup>), and similarly for β<sub>3</sub>.

Or, if you need only _I_<sub>1</sub> and _I_<sub>3</sub>, then analyze the difference image using `order=3`, and take `_, I1, _, I3 = ...harmonics()`, ignoring _I_<sub>0</sub> and _I_<sub>2</sub>, since they should be zero for purely antisymmetric distributions (although I would suggest to look at them and confirm that they actually are zero, as experimental data always have some errors, and if you see that even terms are noticeably non-zero, it would indicate that your results are skewed, and the odd terms would mean not what you expect).

**Comment by haribo3127** [comment]
Again sorry for the delay. I attach here an example of antisymmetric image (i.e IM+ - IM-) that can be tested. I will now try to extract the anisotropy parameters as suggested above. However, another idea is to extract only the odd parameters from the antisymmetric image while extracting the even parameters from the symmetric image (i.e, IM+ + IM-). In this case it would be useful if one can select the orders to be included in the expansion. It would be interesting to compare both methods.
[antisymtest.txt](https://github.com/PyAbel/PyAbel/files/3831540/antisymtest.txt)

**Comment by MikhailRyazanov** [comment]
Thanks for the example!

Some remarks:
1. It looks very strange: totally symmetric horizontally and totally antisymmetric vertically (except the border), and also very much blurred.
2. PyAbel assumes that the origin is in the center of the central pixel, but your image is 210×210 and has symmetry axes between the pixels. So I had to downscale it by a factor of 2 (it is blurry anyway) to 105×105.
3. The data does not really look like a 3rd-order distribution (I assume that it is already Abel-inverted, but even if not, Abel transform only mixes existing orders, without adding higher-order terms), it is either extremely noisy or has higher harmonics.

In any case, here is its analysis:
```python
import numpy as np
from scipy.ndimage import zoom
import matplotlib.pyplot as plt
from abel.tools.vmi import harmonics

IM = np.loadtxt('antisymtest.txt')
IM = zoom(IM, 0.5, order=1)

fig = plt.figure(figsize=(16, 8))

fig.add_subplot(1, 2, 1)
plt.imshow(IM, cmap='bwr')

fig.add_subplot(2, 2, 2)
for n, c in enumerate(harmonics(IM, order=3)):
  plt.plot(c, label='$P_{}(\\cos\\theta)$'.format(n))
plt.legend()

fig.add_subplot(2, 2, 4)
for n, h in list(enumerate(harmonics(IM, order=7)))[1::2]:
  plt.plot(h, label='$P_{}(\\cos\\theta)$'.format(n))
plt.legend()
plt.xlabel('$r$')

plt.tight_layout()
plt.show()
```
![antisymtest](https://user-images.githubusercontent.com/12854133/68723855-17fe2480-0577-11ea-9533-8c6f042a4039.png)

As you can see from the first plot (upper right), the even harmonics are zero everywhere (since your image is totally antisymmetric), so you can simply ignore them, as I have done in the second plot (lower right). And the second plot shows that analysis with more orders gives the same 1st and 3rd harmonics (as it should, since all harmonics are mutually orthogonal), although the higher harmonics here are not much lower in intensity, indicating that either the data is too noisy or the assumptions are wrong.

Regarding analyzing IM<sup>+</sup> ± IM<sup>−</sup> or IM<sup>±</sup>, everything in the calculations (except anisotropy parameters β, which anyway cannot by obtained directly from IM<sup>+</sup> − IM<sup>−</sup>) is linear, so the results must be the same. As I've shown above, you can do `harmonics(...)[1::2]` to get only the odd terms.

In principle, doing calculations for only odd terms would be more efficient (and more robust in pathological cases), and I, also in principle, can add this capability, and even a capability to specify an arbitrary list of orders, but either of these options will make the code significantly more complicated. So unless somebody really-really needs it for real-time processing, I would prefer not to do that, at least now.

**Comment by MikhailRyazanov** [comment]
Here are, by the way, the images reconstructed from these harmonics up to 3rd and 7th order:
![test37](https://user-images.githubusercontent.com/12854133/68732024-3ec85500-058f-11ea-8d15-284477f69c1b.png)
```python
...

def image(distr, IM):
  c = distr(IM).cos()[1::2]
  cos = distr.c[1::2]
  rbin = distr.bin
  im = np.sum((distr.wl * np.append(c[i], [0])[rbin] * cos[i] +
               distr.wu * np.append(c[i][1:], [0, 0])[rbin] * cos[i]
               for i in range(len(c))))
  return np.hstack((im[:, :0:-1], im))

fig.add_subplot(1, 2, 1)
plt.title('Up to 3rd order')
plt.imshow(image(Distributions(order=3), IM), cmap='bwr')

fig.add_subplot(1, 2, 2)
plt.title('Up to 7th order')
plt.imshow(image(Distributions(order=7), IM), cmap='bwr')

...
```
And their differences from the original image (`IM - image(...)`) on the same intensity scale:
![diff37](https://user-images.githubusercontent.com/12854133/68732657-6ddfc600-0591-11ea-99af-408c8b7f485b.png)

**Comment by MikhailRyazanov** [comment]
@haribo3127, do you have any updates on this issue?

**Comment by DanHickstein** [comment]
I finally got a chance to catch up and read through the discussion here. Great work @MikhailRyazanov! It's great that we can now support odd orders for the anisotropy distributions. I've seen lots of interesting papers recently about chiral molecules ionized using circularly polarized light, so I think that this is indeed something that people will use.

**Comment by MikhailRyazanov** [comment]
@DanHickstein, so since you've seen them, can you tell whether they all indeed use plain Legendre polynomials, or there is some other (maybe more useful) formalism? And can we get any real-world examples for testing? ;–)

**Comment by MikhailRyazanov** [comment]
Odd orders are now implemented in #266, so I'm closing this. If somebody has additional considerations, please reopen issue or open a new, more specific one.

**Comment by DanHickstein** [comment]
Awesome. Thanks Mikhail!
